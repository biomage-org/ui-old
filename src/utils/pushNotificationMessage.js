import { message } from 'antd';

const pushNotificationMessage = (type, text, duration = 4) => {
  switch (type) {
    case 'success':
      message.success(text, duration);
      break;
    case 'error':
      message.error(text, duration);
      break;
    case 'info':
      message.info(text, duration);
      break;
    case 'warning':
    case 'warn':
      message.warn(text, duration);
      break;
    case 'loading':
      message.loading(text, duration);
      break;
    default:
      message.info(text, duration);
      break;
  }
};

export default pushNotificationMessage;
