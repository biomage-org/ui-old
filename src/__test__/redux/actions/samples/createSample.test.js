import _ from 'lodash';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import fetchMock, { enableFetchMocks } from 'jest-fetch-mock';
import { v4 as uuidv4 } from 'uuid';

import createSample from 'redux/actions/samples/createSample';
import initialSampleState from 'redux/reducers/samples/initialState';
import initialExperimentState, { experimentTemplate } from 'redux/reducers/experiments/initialState';

import {
  SAMPLES_CREATE, SAMPLES_SAVING, SAMPLES_ERROR, SAMPLES_SAVED,
} from 'redux/actionTypes/samples';

import endUserMessages from 'utils/endUserMessages';
import pushNotificationMessage from 'utils/pushNotificationMessage';

pushNotificationMessage.mockImplementation(() => async () => { });

enableFetchMocks();

const mockStore = configureStore([thunk]);

jest.mock('uuid');
const sampleUuid = 'abc123';
uuidv4.mockImplementation(() => sampleUuid);

const sampleName = 'test sample';

describe('createSample action', () => {
  const experimentId = 'exp234';

  const mockType = '10X Chromium';

  const mockExperiment = {
    ...experimentTemplate,
    name: 'Experiment 1',
    id: experimentId,
  };

  const initialState = {
    samples: { ...initialSampleState },
    experiments: {
      ...initialExperimentState,
      [experimentId]: mockExperiment,
      ids: [mockExperiment.id],
    },
  };

  let store;

  beforeEach(() => {
    jest.clearAllMocks();
    jest.useFakeTimers('modern').setSystemTime(new Date('2020-01-01').getTime());

    fetchMock.resetMocks();
    fetchMock.doMock();

    store = mockStore(initialState);
  });

  it('Works correctly with one file being uploaded', async () => {
    fetchMock.mockResponse(JSON.stringify({}), { url: 'mockedUrl', status: 200 });

    const newUuid = await store.dispatch(createSample(experimentId, sampleName, mockType, ['matrix.tsv.gz']));

    // Returns a new sampleUuid
    expect(newUuid).toEqual(sampleUuid);

    // Fetch call is made
    const fetchMockFirstCall = fetchMock.mock.calls[0];

    const { body: fetchBody, method: fetchMethod } = fetchMockFirstCall[1];
    expect(fetchMockFirstCall[0]).toEqual(`http://localhost:3000/v2/experiments/${mockExperiment.id}/samples/${sampleUuid}`);

    expect(fetchMethod).toEqual('POST');
    expect(JSON.parse(fetchBody)).toMatchSnapshot();

    // Sends correct actions
    const actions = store.getActions();
    expect(_.map(actions, 'type')).toEqual([SAMPLES_SAVING, SAMPLES_CREATE, SAMPLES_SAVED]);
    expect(_.map(actions, 'payload')).toMatchSnapshot();
  });

  it('Works correctly with many files being uploaded', async () => {
    fetchMock.mockResponse(JSON.stringify({}), { url: 'mockedUrl', status: 200 });

    const newUuid = await store.dispatch(createSample(experimentId, sampleName, mockType, ['matrix.tsv.gz', 'features.tsv.gz', 'barcodes.tsv.gz']));

    // Returns a new sampleUuid
    expect(newUuid).toEqual(sampleUuid);

    // Fetch call is made
    const fetchMockFirstCall = fetchMock.mock.calls[0];

    const { body: fetchBody, method: fetchMethod } = fetchMockFirstCall[1];
    expect(fetchMockFirstCall[0]).toEqual(`http://localhost:3000/v2/experiments/${mockExperiment.id}/samples/${sampleUuid}`);

    expect(fetchMethod).toEqual('POST');
    expect(JSON.parse(fetchBody)).toMatchSnapshot();

    // Sends correct actions
    const actions = store.getActions();
    expect(_.map(actions, 'type')).toEqual([SAMPLES_SAVING, SAMPLES_CREATE, SAMPLES_SAVED]);
    expect(_.map(actions, 'payload')).toMatchSnapshot();
  });

  it('Throws if the api fails', async () => {
    fetchMock.mockRejectOnce(() => Promise.reject(new Error('Some error')));

    await expect(
      store.dispatch(
        createSample(experimentId, sampleName, mockType, ['matrix.tsv.gz']),
      ),
    ).rejects.toThrow(endUserMessages.ERROR_CREATING_SAMPLE);

    // Sends correct actions
    const actions = store.getActions();
    expect(_.map(actions, 'type')).toEqual([SAMPLES_SAVING, SAMPLES_ERROR]);
    expect(_.map(actions, 'payload')).toMatchSnapshot();
  });

  it('Throws if technology is not identified', async () => {
    fetchMock.mockResponse(JSON.stringify({}), { url: 'mockedUrl', status: 200 });

    await expect(
      store.dispatch(
        createSample(experimentId, sampleName, 'unrecognizable type', ['matrix.tsv.gz', 'features.tsv.gz', 'barcodes.tsv.gz']),
      ),
    ).rejects.toThrow('Sample technology unrecognizable type is not recognized');
  });
});
