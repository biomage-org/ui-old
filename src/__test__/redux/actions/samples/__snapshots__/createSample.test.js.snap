// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createSample action Throws if the api fails 1`] = `
Array [
  Object {
    "message": "Saving sample...",
  },
  Object {
    "error": "We couldn't create your sample. Please try uploading it again",
  },
]
`;

exports[`createSample action Works correctly with many files being uploaded 1`] = `
Object {
  "name": "test sample",
  "sampleTechnology": "10x",
}
`;

exports[`createSample action Works correctly with many files being uploaded 2`] = `
Array [
  Object {
    "message": "Saving sample...",
  },
  Object {
    "experimentId": "exp234",
    "sample": Object {
      "complete": false,
      "createdDate": "2020-01-01T00:00:00.000Z",
      "error": false,
      "fileNames": Array [],
      "files": Object {
        "barcodes.tsv.gz": Object {
          "upload": Object {
            "status": "uploading",
          },
        },
        "features.tsv.gz": Object {
          "upload": Object {
            "status": "uploading",
          },
        },
        "matrix.tsv.gz": Object {
          "upload": Object {
            "status": "uploading",
          },
        },
      },
      "lastModified": "2020-01-01T00:00:00.000Z",
      "metadata": Object {},
      "name": "test sample",
      "projectUuid": "qwe234",
      "type": "10X Chromium",
      "uuid": "abc123",
    },
  },
  Object {
    "experimentId": "exp234",
    "sample": Object {
      "complete": false,
      "createdDate": "2020-01-01T00:00:00.000Z",
      "error": false,
      "fileNames": Array [],
      "files": Object {
        "barcodes.tsv.gz": Object {
          "upload": Object {
            "status": "uploading",
          },
        },
        "features.tsv.gz": Object {
          "upload": Object {
            "status": "uploading",
          },
        },
        "matrix.tsv.gz": Object {
          "upload": Object {
            "status": "uploading",
          },
        },
      },
      "lastModified": "2020-01-01T00:00:00.000Z",
      "metadata": Object {},
      "name": "test sample",
      "projectUuid": "qwe234",
      "type": "10X Chromium",
      "uuid": "abc123",
    },
  },
]
`;

exports[`createSample action Works correctly with one file being uploaded 1`] = `
Object {
  "name": "test sample",
  "sampleTechnology": "10x",
}
`;

exports[`createSample action Works correctly with one file being uploaded 2`] = `
Array [
  Object {
    "message": "Saving sample...",
  },
  Object {
    "experimentId": "exp234",
    "sample": Object {
      "complete": false,
      "createdDate": "2020-01-01T00:00:00.000Z",
      "error": false,
      "fileNames": Array [],
      "files": Object {
        "matrix.tsv.gz": Object {
          "upload": Object {
            "status": "uploading",
          },
        },
      },
      "lastModified": "2020-01-01T00:00:00.000Z",
      "metadata": Object {},
      "name": "test sample",
      "projectUuid": "qwe234",
      "type": "10X Chromium",
      "uuid": "abc123",
    },
  },
  Object {
    "experimentId": "exp234",
    "sample": Object {
      "complete": false,
      "createdDate": "2020-01-01T00:00:00.000Z",
      "error": false,
      "fileNames": Array [],
      "files": Object {
        "matrix.tsv.gz": Object {
          "upload": Object {
            "status": "uploading",
          },
        },
      },
      "lastModified": "2020-01-01T00:00:00.000Z",
      "metadata": Object {},
      "name": "test sample",
      "projectUuid": "qwe234",
      "type": "10X Chromium",
      "uuid": "abc123",
    },
  },
]
`;
